drivers/platform/maxim/common/maxim_pwm.c

```diff
@@ -0,0 +1,355 @@
+
+	return 0;
+}
+
+/**
+ * @brief Set the PWM duty cycle.
+ * @param desc - Pointer where the configured instance is stored.
+ * @param duty_cycle_ns - PWM duty cycle.
+ * @return 0 in case of success, negative error code otherwise.
+ */
+int32_t max_pwm_set_duty_cycle(struct no_os_pwm_desc *desc,
+			       uint32_t duty_cycle_ns)
+{
+	uint32_t duty_ticks, div;
+	mxc_tmr_pres_t prescaler;
+	mxc_tmr_cfg_t *cfg;
+	int32_t ret;
+
+	div = PeripheralClock / NO_OS_DIV_ROUND_CLOSEST_ULL(NANO, duty_cycle_ns);
+	ret = _get_prescaler(div, &prescaler);
+	if (ret)
+		return ret;
+
+	duty_ticks = MXC_TMR_GetPeriod(MXC_TMR_GET_TMR(desc->id),
+				       MXC_TMR_APB_CLK,
+				       MAX_PWM_PRESCALER_VAL(prescaler),
+				       NO_OS_DIV_ROUND_CLOSEST_ULL(NANO, duty_cycle_ns));
+
+	ret = MXC_TMR_SetPWM(MXC_TMR_GET_TMR(desc->id), duty_ticks);
+	if (ret)
+		return ret;
+
+	cfg = desc->extra;
+	cfg->pres = prescaler;
+	desc->extra = cfg;
+	desc->duty_cycle_ns = duty_cycle_ns;
+
+	return 0;
+}
+
+/**
+ * @brief Get the PWM duty cycle.
+ * @param desc - Pointer where the configured instance is stored.
+ * @param duty_cycle_ns - PWM duty cycle.
+ * @return 0 in case of success, negative error code otherwise.
+ */
+int32_t max_pwm_get_duty_cycle(struct no_os_pwm_desc *desc,
+			       uint32_t *duty_cycle_ns)
+{
+	mxc_tmr_pres_t prescaler;
+	mxc_tmr_cfg_t *cfg;
+	mxc_tmr_regs_t *tmr;
+	uint32_t duty_ticks;
+
+	cfg = desc->extra;
+	prescaler = cfg->pres;
+
+	tmr = MXC_TMR_GET_TMR(desc->id);
+
+	/* Return actual measured value of the duty cycle. */
+	*duty_cycle_ns = NO_OS_DIV_ROUND_CLOSEST_ULL(NANO * tmr->pwm *
+			 MAX_PWM_PRESCALER_VAL(prescaler), PeripheralClock);
+
+	return 0;
+}
+
+/**
+ * @brief Configure the VDDIO level for the PWM pins
+ * @param id - the identification number.
+ * @param vssel - the VDDIO level.
+ * @return 0 in case of success, -EINVAL otherwise.
+ */
+static int32_t _max_pwm_pins_vddio(uint32_t id, mxc_gpio_vssel_t vssel)
+{
+	mxc_gpio_cfg_t pwm_pins;
+
+	switch (id) {
+	case 0:
+		pwm_pins = gpio_cfg_tmr0;
+		break;
+	case 1:
+		pwm_pins = gpio_cfg_tmr1;
+		break;
+	case 2:
+		pwm_pins = gpio_cfg_tmr2;
+		break;
+	case 3:
+		pwm_pins = gpio_cfg_tmr3;
+		break;
+	default:
+		return -EINVAL;
+	}
+
+	pwm_pins.vssel = vssel;
+
+	return MXC_GPIO_Config(&pwm_pins);
+}
+
+/**
+ * @brief Initialize the PWM.
+ * @param desc - PWM descriptor.
+ * @param param - The structure containing PWM init parameters.
+ * @return 0 in case of success, negative error code otherwise.
+ */
+int32_t max_pwm_init(struct no_os_pwm_desc **desc,
+		     const struct no_os_pwm_init_param *param)
+{
+	struct max_pwm_init_param *extra_init;
+	struct no_os_pwm_desc *descriptor;
+	mxc_tmr_pres_t prescaler;
+	mxc_tmr_cfg_t *cfg;
+	uint32_t div;
+	int32_t ret;
+
+	descriptor = (struct no_os_pwm_desc *)no_os_calloc(1, sizeof(*descriptor));
+	if (!descriptor)
+		return -ENOMEM;
+
+	cfg = (mxc_tmr_cfg_t *)no_os_calloc(1, sizeof(*cfg));
+	if (!cfg) {
+		return -ENOMEM;
```

Comment:
this return should be dropped to fix the cppcheck build. the return is handled already on next line with the conditional jump

Reply:
I am guessing just assigning ret = -ENOMEM, so the correct error code is returned.